### PR TITLE
Fix shape inference issue in Gather op

### PR DESCRIPTION
**Pull request by stevenlix**
When Gather has 1D data input, the sympy compute might fail if Gather indices value is not available. Check indices value before the computation

### CODE REVIEW

**Comment by tianleiwu** [comment]
It would be nice to add a test case that having such issue.

**Comment by ke1337** [comment]
> It would be nice to add a test case that having such issue.

Please consider to add tests here: https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/python/onnxruntime_test_python_symbolic_shape_infer.py

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 1]
```diff
@@ -892,7 +892,8 @@ def _infer_Gather(self, node):
             helper.make_tensor_value_info(node.output[0], self.known_vi_[node.input[0]].type.tensor_type.elem_type,
                                           data_shape[:axis] + indices_shape + data_shape[axis + 1:]))
         # for 1D input, do some sympy compute
-        if node.input[0] in self.sympy_data_ and len(data_shape) == 1 and 0 == get_attribute(node, 'axis', 0):
+        if node.input[0] in self.sympy_data_ and (node.input[1] in self.sympy_data_ or node.input[1] in self.initializers_) \
+           and len(data_shape) == 1 and 0 == get_attribute(node, 'axis', 0):
             idx = self._get_value(node, 1)
```
Comment (ke1337) [inline onnxruntime/python/tools/symbolic_shape_infer.py:897]: A better fix might be here: change ```_get_value``` to ```_try_get_value```, which returns None if input[1] is unknown

Comment (stevenlix) [reply onnxruntime/python/tools/symbolic_shape_infer.py:897]: Thanks. Will do

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 2]
```diff
@@ -893,16 +893,18 @@ def _infer_Gather(self, node):
...
         if node.input[0] in self.sympy_data_ and len(data_shape) == 1 and 0 == get_attribute(node, 'axis', 0):
-            idx = self._get_value(node, 1)
-            data = self.sympy_data_[node.input[0]]
-            if type(data) == list:
-                if type(idx) == np.ndarray and len(idx.shape) == 1:
-                    self.sympy_data_[node.output[0]] = [data[int(i)] for i in idx]
+            idx = self._try_get_value(node, 1)
+            if idx is not None:
+                data = self.sympy_data_[node.input[0]]
+                print('data: ', data, ', type(data): ', type(data))
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:899]: remove print that is not useful for customer.

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 3]
```diff
@@ -893,16 +893,18 @@ def _infer_Gather(self, node):
...
+                    if type(idx) == np.ndarray and len(idx.shape) == 1:
+                        self.sympy_data_[node.output[0]] = [data[int(i)] for i in idx]
+                    else:
+                        self.sympy_data_[node.output[0]] = data[int(idx)]
                 else:
-                    self.sympy_data_[node.output[0]] = data[int(idx)]
-            else:
-                assert idx == 0
-                self.sympy_data_[node.output[0]] = data
+                    assert idx == 0
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:906]: The indices be -1 that makes the assertion failure.

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 4]
```diff
@@ -893,16 +893,18 @@ def _infer_Gather(self, node):
...
+            idx = self._try_get_value(node, 1)
+            if idx is not None:
+                data = self.sympy_data_[node.input[0]]
+                print('data: ', data, ', type(data): ', type(data))
+                if type(data) == list:
+                    if type(idx) == np.ndarray and len(idx.shape) == 1:
+                        self.sympy_data_[node.output[0]] = [data[int(i)] for i in idx]
+                    else:
+                        self.sympy_data_[node.output[0]] = data[int(idx)]
                 else:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:905]: The [onnx spec](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Gather) mentioned that the indices is "Tensor of int32/int64 indices". what situation that it is not list?

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**